Source/WebDriver/WebDriverService.cpp

```diff
@@ -921,6 +1036,35 @@ void WebDriverService::createSession(Vector<Capabilities>&& capabilitiesList, st
             timeoutsObject->setDouble("implicit"_s, m_session->implicitWaitTimeout());
             capabilitiesObject->setObject("timeouts"_s, WTFMove(timeoutsObject));
 
+#if ENABLE(WEBDRIVER_BIDI)
+            // Extension steps defined by BiDi spec: https://w3c.github.io/webdriver-bidi/#establishing
+            // 1. If flags contains "bidi", return.
+            // 2. Let webSocketUrl be the result of getting a property named "webSocketUrl" from capabilities.
+            // 3. If webSocketUrl is undefined or false, return.
+            // 4. Assert: webSocketUrl is true.
+            if (!m_session->bidiFlag() && capabilities.webSocketURL && *capabilities.webSocketURL) {
+                // 5. Let listener be the result of start listening for a WebSocket connection given session.
+                auto listener = m_bidiServer.startListeningForAWebSocketConnectionGivenSession(m_session->id());
+                // Implementation note: We need to update the listener host to a visible one so remote clients can connect to it
+                listener->host = m_server.visibleHost();
+
+                // 6. Set webSocketUrl to the result of constructing a WebSocket URL given listener and session.
+                auto webSocketURL = m_bidiServer.constructWebSocketURLForListenerAndSession(listener, m_session->id());
+                // 7. Set a property on capabilities named "webSocketUrl" to webSocketUrl.
+                capabilitiesObject->setString("webSocketUrl"_s, webSocketURL);
+                // 8. Set session’s BiDi flag to true.
+                m_session->setBiDiFlag(true);
+                // 9. Append "bidi" to flags.
+                // TODO
```

Comment:
?

Reply:
Removed (we don't have an explicit flags member).